Add ability to cancel model loading

crasm opened the pull request:
Updated llama_progress_callback so that if it returns false, the model loading is aborted. This is primarily useful for anybody creating UIs that rely on llama.cpp and let users load models on demand.

The test's success condition is getting a nullptr back from llama_load_model_from_file, so it also passes if the model file is missing. Not sure if the test is worth including, to be honest.

crasm committed:
llama : Add ability to cancel model load

Updated llama_progress_callback so that if it returns false, the model
loading is aborted.

crasm committed:
llama : Add test for model load cancellation

cebtenzzre commented on tests/test-model-load-cancel.cpp:
```diff
@@ -0,0 +1,17 @@
+    auto params = llama_model_params{};
+    params.use_mmap = false;
+    params.progress_callback = [](float progress, void * ctx){
+        std::ignore = ctx;
```
According to cppreference, "the behavior of std::ignore outside of std::tie is not formally specified."

cebtenzzre commented on llama.cpp:
```diff
@@ -3712,16 +3722,18 @@ static bool llama_model_load(const std::string & fname, llama_model & model, con
             return true;
```
This function is returning true, 0, -1, and -2?

crasm committed:
Fix bool return in llama_model_load, remove std::ignore use

crasm commented:
@cebtenzzre This should be ready. Can you please take another look? Thank you!

ggerganov reviewed (approved):
Merge after @cebtenzzre approves

cebtenzzre commented on llama.cpp:
```diff
@@ -9017,11 +9029,18 @@ struct llama_model * llama_load_model_from_file(
+        if (status == -1) {
+            LLAMA_LOG_ERROR("%s: failed to load model\n", __func__);
+        } else if (status == -2) {
+            LLAMA_LOG_INFO("%s, cancelled model load\n", __func__);
```
```suggestion
            LLAMA_LOG_INFO("%s: cancelled model load\n", __func__);
```
btw should we be using British English in strings or American English?

ggerganov replied:
No preference, cancelled is fine

ggerganov committed:
Update llama.cpp

Co-authored-by: Jared Van Bortel <cebtenzzre@gmail.com>

crasm committed:
Fail test if model file is missing

crasm committed:
Merge branch 'cancel-model-load' of github.com:crasm/llama.cpp into cancel-model-load

crasm commented:
Sorry, had an idea to check for the model file exists. If this fails CI I can remove it

crasm committed:
Revert "Fail test if model file is missing"

This reverts commit 32ebd525bf7e5a87ee8a3dbaab3d92ce79fbf23d.

cebtenzzre commented on tests/test-model-load-cancel.cpp:
This test should be added to the Makefile, and we probably *should* find a way to make it pass with a warning if the model isn't found.

crasm replied:
In that case, how about I create a mock/stub gguf that is just metadata with a few fake tensors, and commit that to the repo? I can try to figure that out if you like the idea.

ggerganov replied:
I think it is easier to add it to https://github.com/ggerganov/llama.cpp/tree/master/ci
In some of the tests, we already download a model and it can be used to run this test on it.

crasm replied:
Hmm. All of the CI parts that use models are in `gg_run_open_llama_{3,7}b_v2`. But before those are run, `gg_run_ctest_{debug,release}` are run. None of the existing ctests require models, and downloading models (and running convert.py on them) before running the other tests would probably be a mistake.

If we want to have ctests that run with models rather than mocking out the gguf for this one test, it'll require creating a separate tests directory or subdirectory (and in the cmake/make), and a separate CI run for those tests that happen after `gg_run_open_llama_{3.7}b_v2`.

So it's not trivial to add this test to the CI. However, it might still be worth it, so that it's easy in the future to add CI tests that need the models.

ggerganov replied:
Ah, yes. If we want to go the "mock gguf" way, you can take a look at the `examples/gguf` example - it has functionality to write a gguf file. I guess we can update the `test-model-load-cancel` to write a small "llama" model and then load it with cancellation. But it can get tedious because you would need to populate the entire model - i.e. all required KV meta + all required tensors. The benefit of this would be that we don't have to commit the mock file - it will be generated by the test on-the-fly.

Another approach which seems easier is to add [ctest label](https://cmake.org/cmake/help/latest/prop_test/LABELS.html) to the tests that require a model (for example, label `model`) and then run a `ctest -L model` command inside `gg_run_open_llama_{3,7}b_v2` where we already have model data. The existing `ctest` commands will be changed to run `ctest -LE model` so that they exclude the tests that depend on model data

crasm committed:
Add test-model-load-cancel to Makefile

crasm committed:
Revert "Revert "Fail test if model file is missing""

This reverts commit 2796953257ee5383fa7c8fe8fa8fc888c048fb0b.

crasm committed:
Simplify .gitignore for tests, clang-tidy fixes

crasm committed:
Label all ctest tests

crasm committed:
ci : ctest uses -L main

crasm committed:
Attempt at writing ctest_with_model

crasm committed:
ci : get ci/run.sh working with test-model-load-cancel

crasm committed:
ci : restrict .github/workflows/build.yml ctest to -L main

crasm committed:
update requirements.txt

crasm commented:
The ctest labels are working great for cmake. However, it's failing on `make test` in `build.yml` and `code-coverage.yml` because there's no model file on the github workflow.

* Should I find a way to selectively run this test in make if it's local vs in the CI/CD pipeline?
* Change the test name and do some globbing? (idk if possible in make)
* Create a new target separate from `make test`? So we'd have `BUILD_TARGETS`, `TEST_TARGETS`, and `TEST_WITH_MODEL_TARGETS`
* ...?

I could also use some feedback on the direction I'm taking the `run.sh` script. I would like to make it more robust, cut down on the repetitive logic duplicated among 3B/7B/debug/release, and have a clean shellcheck—not sure I can do all that in this PR, however.

What state should I be aiming for to get this merged?

ggerganov commented on requirements.txt:
```diff
@@ -1,4 +1,6 @@
 numpy==1.24.4
 sentencepiece==0.1.98
+torch>=2.0.0
```
I don't think `torch` is needed

crasm replied:
It otherwise fails with a clean venv, so I needed to add it. I see the note in `convert.py` about dropping it as a dependency, but it is imported in a number of python files.

```
% find . -name '*.py' -not -path './tmp/*' -print0 | xargs -0 -- grep -nFe 'import torch'
./convert-lora-to-ggml.py:12:import torch
./convert-persimmon-to-gguf.py:1:import torch
./convert-hf-to-gguf.py:16:import torch
./examples/llava/llava-surgery.py:4:import torch
./examples/llava/convert-image-encoder-to-gguf.py:5:import torch
```

`./convert-lora-to-ggml.py` is the one that fails CI. I don't see any other python files being called in `run.sh`.

slaren replied:
Generally we have used `requirements.txt` for `convert.py` only. We also have a different `requirements-hf-to-gguf.txt` for `convert-hf-to-gguf.py`. I am not sure what changed, but `convert-lora-to-ggml.py` doesn't fail currently, but it does require pytorch.

crasm replied:
To clarify, `./convert-lora-to-ggml.py` is called in `run.sh`, and `run.sh` reliably fails with an error if torch is not included in the script's venv.

If it is not failing for you, you must already have pytorch installed in your user pip packages? Before my changes, `run.sh` did not use a virtual environment.

Instead of adding pytorch to `requirements.txt`, I can alternatively do a `pip install torch` in `run.sh` if that's preferable.

ggerganov reviewed (approved):
The PR is already quite useful, so we should merge it.

Let's for now disable the test target in the Makefile and figure out what is the best way to enable it later - I'm not sure as well.

I like the changes in `ci/run.sh`. The only constraint for this script is to not exceed some runtime - currently the limit is 30 minutes: 

https://github.com/ggml-org/ci/blob/master/env.sh#L113

We can increase it if necessary, but only as a last resort.

Aside from refactoring and shellcheck, the script can be improved with more simple tests that rely on model data and are not suitable for Github Actions - i.e. require GPU for example. Any improvements in this direction would be great.

crasm committed:
Disable test-model-load-cancel in make

crasm committed:
Remove venv before creation

crasm committed:
Restructure requirements.txt

Top-level now imports the specific additional requirements for each
python file. Using `pip install -r requirements.txt` will fail if
versions become mismatched in the per-file requirements.

crasm committed:
Merge commit 'c50e40016394f124b97ce39da48148b1f6c01833' into cancel-model-load

crasm committed:
Make per-python-script requirements work alone

This doesn't break the main requirements.txt.

crasm committed:
Add comment

crasm committed:
Merge remote-tracking branch 'upstream/master' into cancel-model-load

crasm commented:
The new organization for `requirements.txt` will force versions across top-level scripts to remain compatible. This way, including torch as a dependency in both `convert-lora-to-ggml.py` and `convert-hf-to-gguf.py` won't end up with weird mismatches where they need different versions of pytorch.

If there are no objections to the new `requirements.txt` structure or any other notes, I will merge this PR tomorrow.

Btw I am getting test failure locally after updating from master. I don't think it's anything I did because it's unrelated to model loading.

<details>

```
[MUL] NMSE = 0.000000 FAIL
  708/709 tests passed
  Backend Metal: FAIL

1/2 backends passed
FAIL

      Start 19: test-rope
19/19 Test #19: test-rope ...........................   Passed    0.08 sec

95% tests passed, 1 tests failed out of 19

Label Time Summary:
main    =  20.27 sec*proc (19 tests)

Total Test time (real) =  20.28 sec

The following tests FAILED:
         18 - test-backend-ops (Failed)
```

</details>

crasm commented:
Failing test from re-running `run.sh` without manually cleaning out the artifacts. It shouldn't be necessary to manually clean things up on re-runs, so I'll look into that as I begin revamping the rest of the script.

@cebtenzzre this is ready to go, pending any objections

crasm committed:
Add convert-persimmon-to-gguf.py to new requirements.txt scheme

ggerganov commented on convert_requirements.txt:
Let's rename all `convert-*_requirements.txt` to `requirements-convert-*.txt`

Note to use dash instead of underscore

nittaya111 commented:
😎 what? My family 😎

เมื่อ พ. 20 ธ.ค. 2566 เวลา 15:40 Georgi Gerganov ***@***.***>
เขียนว่า:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> On convert_requirements.txt
> <https://github.com/ggerganov/llama.cpp/pull/4462#discussion_r1432413725>:
>
> Let's rename all convert-*_requirements.txt to requirements-convert-*.txt
>
> Note to use dash instead of underscore
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/ggerganov/llama.cpp/pull/4462#pullrequestreview-1790418277>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/BDWM7ZX2VBIHFZQCG27RBRDYKKQBDAVCNFSM6AAAAABAURRMXSVHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMYTOOJQGQYTQMRXG4>
> .
> You are receiving this because you are subscribed to this thread.Message
> ID: ***@***.***>
>

LoganDark commented:
https://github.com/ggerganov/llama.cpp/blob/799fc2268989482054944c902874cca76337580f/ggml-cuda.cu#L9094

now needs to check that `tensor->extra` is not NULL

![image](https://github.com/ggerganov/llama.cpp/assets/4723091/9cb700f2-60d8-4490-b137-d8f86a45ee84)

```diff
-    if (!tensor || (tensor->backend != GGML_BACKEND_GPU && tensor->backend != GGML_BACKEND_GPU_SPLIT) ) {
+    if (!tensor || !tensor->extra || (tensor->backend != GGML_BACKEND_GPU && tensor->backend != GGML_BACKEND_GPU_SPLIT) ) {
```

fixes the access violation

crasm committed:
Add check-requirements.sh script and GitHub workflow

crasm committed:
Remove shellcheck installation step from workflow

crasm committed:
Add nocleanup special arg

crasm committed:
Merge remote-tracking branch 'upstream' into cancel-model-load

slaren commented on llama.cpp:
```diff
@@ -3722,16 +3729,23 @@ static void llm_load_tensors(
+    if (progress_callback) {
+        // Even though the model is done loading, we still honor
+        // cancellation since we need to free allocations.
+        ok = ok && progress_callback(1.0f, progress_callback_user_data);
```
The last `progress_callback` call is done in `load_all_data` now.

crasm replied:
I missed that. However, that's slightly different behavior. I wanted to make sure that if the loading is aborted even at `1.0f` that it still aborts.

crasm committed:
Fix merge

see: https://github.com/ggerganov/llama.cpp/pull/4462#discussion_r1434593573

crasm commented on llama.cpp:
```diff
@@ -3729,19 +3729,16 @@ static bool llm_load_tensors(
+        return false;
     }
 
     model.mapping = std::move(ml.mapping);
```
@slaren do you know if this line will be a problem? Since it doesn't get run if the above returns early

LoganDark replied:
> @slaren do you know if this line will be a problem? Since it doesn't get run if the above returns early

The progress callback should only be called if it loaded successfully, I think. Would be weird to run it with 1.0 if the model load actually failed

slaren replied:
Skipping the `mapping` move should be fine.

crasm replied:
The case I'm trying to **avoid** is:
1. The model has fully loaded
2. The client has received some information that the model load should be cancelled (e.g. on another thread)
3. llama.cpp calls the progress_callback with `1.0f`
4. Client returns false (independent from the `1.0f` value itself)
5. llama.cpp ignores the callback return value, and returns a valid `llama_model *`
6. Now the client has to free the model, even though it did not expect one to be returned

crasm commented:
Would greatly appreciate this is ok'd for merge sooner rather than later. I am not that confident I can fix merge conflicts without missing things

slaren commented:
I tried running the CI locally, but the `test-model-load-cancel` fails due to not finding the model. Is it supposed to work in the CI with the open llama models?

```
    Start 20: test-model-load-cancel
1/1 Test #20: test-model-load-cancel ...........***Failed    0.02 sec
no model found
```

crasm commented:
> I tried running the CI locally, but the `test-model-load-cancel` fails due to not finding the model. Is it supposed to work in the CI with the open llama models?

It's supposed to. The test itself looks for the Q2_k quant for 3B and 7B in the models-mnt directory.

I'm assuming your `models-mnt/open-llama/7B-v2` directory exists?

I'll try to debug, but it'll be a little tricky because I don't have an Nvidia device operational at the moment (so the compile fails).

slaren commented:
Yes, but the current directory doesn't seem to be what the test expects.

```
+ gg_run_ctest_with_model
+ cd /home/diego/code/llama.cpp
+ tee /home/diego/code/llama.cpp/tmp/results/ctest_with_model.log
+ cd build-ci-release
+ set -e
+ tee -a /home/diego/code/llama.cpp/tmp/results/ctest_with_model-ctest_with_model.log
+ ctest --output-on-failure -L model
Test project /home/diego/code/llama.cpp/build-ci-release
    Start 20: test-model-load-cancel
1/1 Test #20: test-model-load-cancel ...........***Failed    0.02 sec
```

The models are in `/home/diego/code/llama.cpp/tmp/mnt/models/open-llama/7B-v2/`

slaren commented:
This path works: `"../../models-mnt/open-llama/3B-v2/ggml-model-q2_k.gguf"`.

slaren commented on tests/test-model-load-cancel.cpp:
```diff
@@ -0,0 +1,46 @@
+        "./models-mnt/open-llama/7B-v2/ggml-model-q2_k.gguf",
+    };
+
+    const char * chosen_model;
```
```suggestion
    const char * chosen_model = nullptr;
```

crasm commented:
Hmm, I need to go over the CI script again. I seem to be unable to get the test to fail at all, even when I sabotage it—though it fails and succeeds as expected when run manually. May be related to the path issue.

crasm commented:
I'm going to split this PR up. The precipitating feature (cancelling model load) is nearly trivial and can be individually reviewed and merged.

It's the consequences of integrating the test that are overcomplicating this.

This PR will be updated to contain only the changes to `llama.h` and `llama.cpp`, without the test.

The other PRs will be:
* Requirements.txt checking and updates
* Revamp of `ci/run.sh` and model load test

crasm committed:
reset to upstream/master

crasm committed:
Redo changes for cancelling model load
